src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -963,13 +959,6 @@ intptr_t ObjectSynchronizer::FastHashCode(Thread* current, oop obj) {
       }
       // Fall thru so we only have one place that installs the hash in
       // the ObjectMonitor.
-    } else if (LockingMode == LM_LIGHTWEIGHT && mark.is_fast_locked() && is_lock_owned(current, obj)) {
```

Comment:
You can delete the is_lock_owned() function just above this with this patch.

Reply:
Also I think the comment before this function is wrong.  We can inflate for deoptimization.

Reply:
> but you could move it down to before 'inflate'
Done.

> Also I think the comment before this function is wrong. We can inflate for deoptimization.
The comment on `is_lock_owned`? For relock_objects `current` is the deoptee's thread which is a JavaThread.

Reply:
Yes that comment.  If we don't inflate for hashcode anymore, why would we inflate with a non-JavaThread? Is it impossible?

Reply:
I see what you mean. Because we need to put the hash code in the ObjectMonitor if it already is inflated and it may be the case that the ObjectMontior has an anonymous owner we can still call `is_lock_owned` when retrieving the monitor for FastHashCode. 

Technically we could condition the inflate call on if the mark word lock bits and pick out the ObjectMontior from the mark word. It would always be installed when reaching that point in the LM_LIGHTWEIGHT mode.

Reply:
LM_LEGACY and LM_MONITOR will. LM_LIGHTWEIGHT technically may. If deflation finishes between reading the mark word in FastHashCode and reading the mark word in `inflate`. It seems like a rare enough case that it does not need to be handled separately. The following change would avoid inflation all together.
```diff
    // An async deflation can race after the inflate() call and before we
    // can update the ObjectMonitor's header with the hash value below.
+   if (LockingMode == LM_LIGHTWEIGHT) {
+     assert(mark.has_monitor(), "must be");
+     monitor =  mark.monitor();
+   } else {
-   monitor = inflate(current, obj, inflate_cause_hash_code);
+     monitor = inflate(current, obj, inflate_cause_hash_code);
+   }
    // Load ObjectMonitor's header/dmw field and see if it has a hash.
```

Maybe I should change it to this. Given that there has been confusion here.
My ideal solution would be to separate the implementations for the different locking modes all together, all of these functions are littered with if (LockingMode == X).

Reply:
So what did you decide to do here?

Reply:
For now I believe the extra code noise from trying to handle this race with deflation is not worth it. I creates some questionable code paths and head scratchers. If we were to add a separate FastHashCode just for LM_LIGHTWEIGHT it would be worth it as the while loop body would look quite a bit different and be easier to reason about. 

But I was looking for input if we should handle this case regardless of code complexity. Or maybe taking this all the way and create a separate FastHashCode with its own more understandable logic which does not have to try to fit in with the legacy locking/inflation protocol.

Regardless if we were to just go with it as it is now there should probably be a comment here along the line:
```c++
  // With LM_LIGHTWEIGHT FastHashCode may race with deflation here and cause a monitor to be re-inflated.
```

Reply:
Okay I've re-read this group of comments multiple times and above you wrote:

> Regardless if we were to just go with it as it is now there should probably be a comment here along the line:

```
>   // With LM_LIGHTWEIGHT FastHashCode may race with deflation here and cause a monitor to be re-inflated.
```

To fully understand whether the comment is correct or not, I need to understand
where the "here" is that you want to place this comment. You deleted old line 972
thru old line 978. Is this new comment going to be a replacement for those lines?
Or will the new comment be somewhere else?

Reply:
Before the call to inflate which is the only control flow path which can observe the race. But it could also be a more general comment overarching the whole FastHashCode. 

The race occurs between reading the mark in FastHashCode and reading the mark in inflate. The ObjectMonitor found in the first mark read may differ from the one returned from inflate, and it may be that the current thread which called FastHashCode is the one that inflates the monitor.

I believe we can leave this enhancement as is, it solves the common case by avoiding inflation when an object is lightweight locked. The race where a re-inflation may occur can be plugged as a separate enhancement.